File: src/main/java/dark/leech/text/plugin/js/api/Html.java
package dark.leech.text.plugin.js.api;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;

import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;
import org.mozilla.javascript.Context;
import org.mozilla.javascript.Scriptable;

import dark.leech.text.action.Log;
import dark.leech.text.util.SettingUtils;

/**
 * HTML API for JavaScript plugins using Rhino. Provides HTML parsing, selection, URL encoding, and
 * element manipulation. Simple Java class - public methods automatically exposed to JavaScript.
 */
public class Html extends JsApiWrapper {

    private static final org.jsoup.nodes.Document EMPTY_DOC =
            org.jsoup.Jsoup.parse("<html><body></body></html>");

    /**
     * Create Html API with execution context for proper object creation.
     *
     * @param context The Rhino Context for script execution
     * @param scope The Rhino Scriptable scope
     */
    public Html(Context context, Scriptable scope) {
        super(context, scope);
    }

    /** Create Html API without context (legacy compatibility). */
    public Html() {
        super();
    }

    /**
     * Parse HTML string into JSDocument wrapper. Usage: html.parse("<html>...</html>") Returns
     * empty document wrapper for null/empty input (vBook compatibility).
     */
    public JSDocument parse(String html) {
        if (html == null || html.isEmpty()) {
            return new JSDocument(EMPTY_DOC);
        }
        org.jsoup.nodes.Document doc = org.jsoup.Jsoup.parse(html);
        return new JSDocument(doc);
    }

    /**
     * Parse HTML from URL. Usage: html.parseUrl("https://example.com") Returns empty document
     * wrapper on failure (vBook compatibility).
     */
    public JSDocument parseUrl(String url) {
        if (url == null || url.isEmpty()) {
            return new JSDocument(EMPTY_DOC);
        }
        try {
            org.jsoup.nodes.Document doc =
                    org.jsoup.Jsoup.connect(url)
                            .userAgent(SettingUtils.USER_AGENT)
                            .timeout(SettingUtils.TIMEOUT)
                            .get();
            return new JSDocument(doc);
        } catch (Exception e) {
            Log.add("Failed to parse HTML from URL: " + e.getMessage());
            return new JSDocument(EMPTY_DOC);
        }
    }

    /** URL encode a string. Usage: html.urlEncode("hello world") => "hello+world" */
    public String urlEncode(String url) {
        return urlEncode(url, StandardCharsets.UTF_8.name());
    }

    public String urlEncode(String url, String charset) {
        try {
            return URLEncoder.encode(url, charset);
        } catch (Exception e) {
            Log.add("URL encoding failed: " + e.getMessage());
            return url;
        }
    }

    /** URL decode a string. */
    public String urlDecode(String url) {
        return urlDecode(url, StandardCharsets.UTF_8.name());
    }

    public String urlDecode(String url, String charset) {
        try {
            return java.net.URLDecoder.decode(url, charset);
        } catch (Exception e) {
            Log.add("URL decoding failed: " + e.getMessage());
            return url;
        }
    }

    /**
     * Clean HTML by removing specified tags. Usage: html.clean("<html>...</html>", ["script",
     * ".ads"])
     */
    public static String clean(String html, Object[] tags) {
        if (html == null || html.isEmpty()) {
            return html;
        }
        org.jsoup.nodes.Document doc = org.jsoup.Jsoup.parse(html);

        if (tags != null) {
            for (Object tag : tags) {
                if (tag != null) {
                    String tagStr = org.mozilla.javascript.Context.toString(tag);
                    if (tagStr != null && !tagStr.isEmpty()) {
                        doc.select(tagStr).remove();
                    }
                }
            }
        }
        return doc.body().html();
    }

    /** Create JSElement wrapper from JSoup Element. */
    public static JSElement wrapElement(Element element) {
        return element != null ? new JSElement(element) : null;
    }

    /** Create JSElements wrapper from JSoup Elements. */
    public static JSElements wrapElements(Elements elements) {
        return new JSElements(elements);
    }

    /** Parse and select in one call. Usage: html.select("<html>...</html>", "div.class") */
    public JSElements select(String html, String selector) {
        JSDocument doc = parse(html);
        return doc.select(selector);
    }
}

File: engine/src/commonMain/kotlin/dev/haipham22/leechtext/plugin/js/api/Html.kt
package dev.haipham22.leechtext.plugin.js.api

import com.fleeksoft.ksoup.Ksoup
import dev.haipham22.leechtext.EngineConfig
import dev.haipham22.leechtext.log.EngineLogger
import dev.haipham22.leechtext.util.decodeBytes
import dev.haipham22.leechtext.util.encodeBytes

/**
 * HTML API cho JS plugins (port từ Html.java; P5.2c common). parse() trả JSDocument;
 * input null/empty → empty document (vBook compat). parseUrl fetch qua HttpEngine seam.
 */
class Html(
    private val log: EngineLogger,
) {
    fun parse(html: String?): JSDocument {
        if (html.isNullOrEmpty()) return JSDocument(EMPTY_DOC)
        return JSDocument(Ksoup.parse(html, ""))
    }

    fun parseUrl(url: String?): JSDocument {
        if (url.isNullOrEmpty()) return JSDocument(EMPTY_DOC)
        return try {
            // Fetch (share HttpEngine) → ksoup parse. baseUri = URL cuối sau redirect cho
            // selector abs:. Non-2xx → EMPTY_DOC như jsoup HttpStatusException cũ
            val result =
                httpExecute(
                    HttpSpec(
                        method = "GET",
                        url = url,
                        headers = mapOf("User-Agent" to EngineConfig.USER_AGENT),
                    ),
                    log,
                )
            if (result.code !in 200..399) {
                error("HTTP ${result.code} từ $url")
            }
            val finalUrl = result.headers["X-Final-Url"] ?: url
            JSDocument(Ksoup.parse(result.body.decodeToString(), finalUrl))
        } catch (e: Exception) {
            log.add("Failed to parse HTML from URL: ${e.message}")
            JSDocument(EMPTY_DOC)
        }
    }

    fun urlEncode(url: String): String = urlEncode(url, "UTF-8")

    fun urlEncode(
        url: String,
        charset: String,
    ): String = try {
        encodeWithCharset(url, charset)
    } catch (e: Exception) {
        log.add("URL encoding failed: ${e.message}")
        url
    }

    fun urlDecode(url: String): String = urlDecode(url, "UTF-8")

    fun urlDecode(
        url: String,
        charset: String,
    ): String = try {
        decodeWithCharset(url, charset)
    } catch (e: Exception) {
        log.add("URL decoding failed: ${e.message}")
        url
    }

    /** Parse + select một lượt: html.select("<html>...</html>", "div.class"). */
    fun select(
        html: String?,
        selector: String,
    ): JSElements = parse(html).select(selector)

    /** Encode form-style theo tên charset (charset không hợp lệ → nguyên input). */
    private fun encodeWithCharset(
        text: String,
        charset: String,
    ): String {
        val bytes = encodeBytes(text, charset) ?: return text
        return urlFormEncode(bytes)
    }

    private fun decodeWithCharset(
        text: String,
        charset: String,
    ): String {
        val bytes = formDecodeToBytes(text)
        return decodeBytes(bytes, charset) ?: text
    }

    companion object {
        private val EMPTY_DOC: com.fleeksoft.ksoup.nodes.Document = Ksoup.parse("<html><body></body></html>", "")

        /** Clean HTML bỏ các tag: Html.clean("<html>...</html>", ["script", ".ads"]). */
        fun clean(
            html: String?,
            tags: Array<Any?>?,
        ): String? {
            if (html.isNullOrEmpty()) return html
            val doc = Ksoup.parse(html, "")
            tags?.asSequence()
                ?.mapNotNull { JsValueBridge.toJsString(it) }
                ?.filter { it.isNotEmpty() }
                ?.forEach { doc.select(it).remove() }
            return doc.body().html()
        }

        fun wrapElement(element: com.fleeksoft.ksoup.nodes.Element?): JSElement? = element?.let { JSElement(it) }

        fun wrapElements(elements: com.fleeksoft.ksoup.select.Elements): JSElements = JSElements(elements)
    }
}
